tools/qflight/qflight.1
.\"                              hey, Emacs:   -*- nroff -*-
.\" qflight is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.\"
.\" See the GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; see the file COPYING.  If not, write to:
.\"
.\"		Free Software Foundation, Inc.
.\"		59 Temple Place, Suite 330
.\"		Boston, MA 02111-1307, USA
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins (default)
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.\"
.TH QFLIGHT 1 "25 August, 2002" QuakeForge "QuakeForge User's Manual"
.\" Please update the above date whenever this man page is modified.
.SH NAME
qflight \- The QuakeForge Light Tool
.SH SYNOPSIS
.B qflight
<\fBcommand\fP> [\fBoptions\fP] \fIFILE\fP...
.SH DESCRIPTION
\fBqflight\fP Takes a .bsp file generated by qfbsp and does light casting from all the static lights in the map.
.SH COMMANDS
\fBqflight\fP accepts the following arguments as a command. Only one command may be
used at a time.
.TP
.B \-h, \-\-help
Show summary of options.
.TP
.B \-V, \-\-version
Show the version of qflight.
.SH OPTIONS
\fBqflight\fP takes the following arguments as options.
.TP
.B \-q, \-\-quiet
Inhibit some of qflight's normal output.
.TP
.B \-v, \-\-verbose
Display more output than usual.
.TP
.B \-t, \-\-threads (n)
The number of threads to use.
.TP
.B \-e, \-\-extra
By default qflight takes one sample from each light point, running with "-extra" enabled takes four samples and averages them. 
This gives smoother lighting/shadows but takes four times as long.
.TP
.B \-d, \-\-distance (n)
Scales the fade distance of ALL lights. n>1 fades faster, n<1 fades slower.  
.TP
.B \-r, \-\-range (n)
Scales the brightness range of ALL lights without affecting their fade distance.
n>.5 makes them brighter, n<.5 makes them darker.
.TP
.SH AUTHORS
Colin Thompson (colinjamesthompson@tiscali.co.uk) ported this to quakeforge from
the original id Software quake utils release.
.SH "SEE ALSO"
.BR quakeforge (1),
.BR qfbsp (1),
.BR qfvis (1)


NEWS
NEWS for the QuakeForge project
-------------------------------

Changes from 0.5.2

    o General enhancements.
        * MVD playback in the qw clients.
        * MVD recording in the qw server.

Changes from 0.3.0

    o General enhancements.
        * NQ (quake1, single player, NetQuake, WinQuake) is back.
        * Video (driver and renderer), audio, gamecode (progs), console,
          model loading and general utility code shared in common libs between
          NQ and QW
        * Plugin system for sound and cd drivers.
        * Ogg Vorbis compression for sound effect samples (streamed ogg support
          is planned).
        * Menu system re-written in QuakeC/Ruamoko
        * Curses console for servers in Unix.
        * New keybinding system.

    o Progs engine enhancements.
        * Fully modular. Mod/Server-independant, allowing for
          stand-alone QuakeC interpreter and client-side code.
        * The engine is no longer dependent on progdefs.h. This makes
          QuakeForge have trouble with QuakeC compilers that strip
          symbol names -- but those can be repaired using the new
          "qfdefs" tool.
        * New basic types, and instructions for using them, for version
          0.fff.002 progs (Quake's Version 6 is still supported).
        * Better dynamic string handling. Dynamic strings are
          garbage-collected.
        * Object-oriented runtime system, in the style of Objective-C.
        * Runtime fixups of built-in functions whose builtin number is
          zero. If a built-in function is available to the engine, with
          the same name given in the source code (e.g.
          "void (void) coredump = #0;"), the engine will set the value
          to the actual number used by the engine.
        * Debugging support (needs compiler support, which qfcc
          provides), including line-number information. Can display the
          line of text in the source on which an error occurred, if the
          new Cvar "pr_source_path" is set correctly.
        * Instruction-level code dumps are now formatted in an
          easier-to-read "assembly-code" format.

    o QuakeForge Code Compiler, qfcc -- a replacement for qcc.
        * Uses libQFutil for general utility functions such as hash
          tables, file access (incomplete), and pak files.
        * Uses libQFgamecode for progs information, such as opcode
          tables, enums, types, etc.)
        * Uses a lex/yacc (actually, Flex and Bison) scanner/parser
          architecture, like a "real" compiler does.
        * Drastically reduces the number of global defs needed for
          a given progs, by reusing temporary globals when possible.
        * All functions use the same block of defs for their local
          variables (compatible with all known servers), which further
          reduces the number of global defs needed.
        * Nested/chained function calls work properly.
        * Simple expressions, such as "a = b + c", compile to a single
          instruction instead of two, and do not use a temporary
          variable.
        * Local variables can be initialized to known values, as in
          "local float foo = 0.1;".
        * Uses the C preprocessor, so header files are now possible and
          useful.
        * The compiler checks for uninitialized and unused variables in
          functions, to help in finding errors.
        * The compiler can generate and use object files, for separate
          compilation (you don't have to compile all of your source all
          the time, only the parts that have changed). Normally, object
          files have a .qfo extension.
        * The compiler can link libraries, which are implemented as .qfo
          files inside pak archives.
        * The compiler implements stricter type checking, to assist in
          finding and fixing code errors.
        * The C language's "for", "do...while", and "switch" control
          structures have been adopted to provide a richer set of
          structures for programmers to use.
        * Additional functionality is given for progs that are to be
          used with QuakeForge's more advanced progs engine:

            x IMPORTANT:
              The order of operations is slightly different in "native"
              mode. The evaluation of "!foo & bar" is no longer
              "!(foo & bar)", but rather "(!foo) & bar". The new order of
              operations rules is more correct, but is different from how
              QCC evaluated. If you give the --traditional switch to
              qfcc's command-line, the old behavior is used.
            x The compiler now allows you to create your own functions
              that can accept a variable number of arguments, using the
              special "..." notation. The special constants "@argc" and
              "@argv" help with this. @argc contains the number of
              arguments, and @argv contains a list of those arguments. The
              handling of this is an improvement on C's vararg function
              handling, in our opinion, made possible by how the Quake
              system works.
            x New basic types: integer, pointer, enum, and id.
            x You can create new complex types, like arrays, structures,
              and unions, and allocate them at runtime.
            x String operators. You can add strings together to
              concatenate them, and you can compare their contents using
              the <=, <, >, >= operators.
            x The integer type has the full array of C operations
              available to it.
            x QFCC and QuakeForge now contain an object-oriented
              programming system, in the style of Objective-C and
              Smalltalk. It's *very* powerful, and can be used to create
              lots of interesting new things with rather little effort. If
              you don't wish to use the new system, you can safely ignore

tools/qflight/source/qflight.c
/*
	trace.c

	(description)

	Copyright (C) 1996-1997  Id Software, Inc.
	Copyright (C) 2002 Colin Thompson

	This program is free software; you can redistribute it and/or
	modify it under the terms of the GNU General Public License
	as published by the Free Software Foundation; either version 2
	of the License, or (at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

	See the GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program; if not, write to:

		Free Software Foundation, Inc.
		59 Temple Place - Suite 330
		Boston, MA  02111-1307, USA

*/
#ifdef HAVE_CONFIG_H
# include "config.h"
#endif

static __attribute__ ((unused)) const char rcsid[] =
	"$Id$";

#ifdef HAVE_UNISTD_H
# include <unistd.h>
#endif
#ifdef HAVE_IO_H
# include <io.h>
#endif
#ifdef HAVE_STRING_H
# include <string.h>
#endif
#ifdef HAVE_STRINGS_H
# include <strings.h>
#endif
#include <getopt.h>
#include <errno.h>
#include <stdlib.h>

#include "QF/bspfile.h"
#include "QF/dstring.h"
#include "QF/mathlib.h"
#include "QF/qtypes.h"
#include "QF/quakefs.h"
#include "QF/sys.h"
#include "QF/va.h"

#include "light.h"
#include "threads.h"
#include "entities.h"
#include "options.h"

options_t	options;
bsp_t *bsp;

char *bspfile;
dstring_t *litfile;

float scalecos = 0.5;

dstring_t *lightdata;
dstring_t *rgblightdata;

dmodel_t *bspmodel;
int bspfileface;		// next surface to dispatch
int bspfileent;			// next entity to dispatch

vec3_t bsp_origin;

qboolean extrasamples;

float minlights[MAX_MAP_FACES];


int
GetFileSpace (int size)
{
	int ofs;

	LOCK;
	lightdata->size = (lightdata->size + 3) & ~3;
	ofs = lightdata->size;
	lightdata->size += size;
	dstring_adjust (lightdata);

	rgblightdata->size = (ofs + size) * 3;
	dstring_adjust (rgblightdata);
	UNLOCK;
	return ofs;
}

static void *
VisThread (void *junk)
{
	int         i;

	while (1) {
		LOCK;
		i = bspfileent++;
		UNLOCK;
		if (i >= num_entities)
			return 0;
		VisEntity (i);
	}
}

static void *
LightThread (void *l)
{
	int         i;

	while (1) {
		LOCK;
		i = bspfileface++;
		UNLOCK;
		if (i >= bsp->numfaces)
			return 0;

		printf ("%5d / %d\r", i, bsp->numfaces);
		fflush (stdout);
		LightFace (l, i);
	}
}

static void
LightWorld (void)
{
	int         i, j;
	vec3_t      org;
	entity_t   *ent;
	const char *name;

	lightdata = dstring_new ();
	rgblightdata = dstring_new ();
	surfacelightchain = (lightchain_t **) calloc (bsp->numfaces,
												  sizeof (lightchain_t *));
	surfaceorgs = (vec3_t *) calloc (bsp->numfaces, sizeof (vec3_t));

	for (i = 1; i < bsp->nummodels; i++) {
		ent = FindEntityWithKeyPair ("model", name = va ("*%d", i));
		VectorZero (org);
		if (!ent)
			Sys_Error ("FindFaceOffsets: Couldn't find entity for model %s.\n",
					   name);
		if (!strncmp (ValueForKey (ent, "classname"), "rotate_", 7))
			GetVectorForKey (ent, "origin", org);
		for (j = 0; j < bsp->models[i].numfaces; j++)
			VectorCopy (org, surfaceorgs[i]);
	}

	VisThread (0);	// not worth threading :/
	VisStats ();
	RunThreadsOn (LightThread);

	BSP_AddLighting (bsp, lightdata->str, lightdata->size);

	if (options.verbosity >= 0)
		printf ("lightdatasize: %i\n", bsp->lightdatasize);
}

int
main (int argc, char **argv)
{
	double      start, stop;
	char       *e;
	QFile      *f;

	start = Sys_DoubleTime ();

	this_program = argv[0];

	DecodeArgs (argc, argv);	

	if (!bspfile) {
		fprintf (stderr, "%s: no bsp file specified.\n", this_program);
		usage (1);
	}

    InitThreads ();

	QFS_StripExtension (bspfile, bspfile);
	QFS_DefaultExtension (bspfile, ".bsp");

	litfile = dstring_new ();
	dstring_copystr (litfile, bspfile);
	e = strrchr (litfile->str, '.');
	dstring_replace (litfile, e - litfile->str, -1, ".lit", 5);

	f = Qopen (bspfile, "rbz");
	bsp = LoadBSPFile (f, Qfilesize (f));
	Qclose (f);
	LoadEntities ();

	MakeTnodes (&bsp->models[0]);

	LightWorld ();

	WriteEntitiesToString ();

	f = Qopen (bspfile, "wb");
	WriteBSPFile (bsp, f);
	Qclose (f);

	dstring_insert (rgblightdata, 0, "QLIT\x01\x00\x00\x00", 8);

	f = Qopen (litfile->str, "wb");
	Qwrite (f, rgblightdata->str, rgblightdata->size);
	Qclose (f);

	stop = Sys_DoubleTime ();
	
	if (options.verbosity >= 0)
		printf ("%5.1f seconds elapsed\n", stop - start);

	return 0;
}
